### PR TITLE
KAFKA-9972: Only commit tasks with valid states

### PR DESCRIPTION
**Pull request by abbccdda**
We spotted a case in the soak test where a standby task could be in `CREATED` state during commit, which causes an illegal state exception. To prevent this from happening, the solution is to always enforce a state check.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Retest this please.

**Review by mjsax — APPROVED** [review]
LGTM (assuming Jenkins passes).

**Comment by abbccdda** [comment]
Hmm, it won't pass as I have already spotted some unit test failures, will fix them.

**Comment by mjsax** [comment]
Retest this please.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:739]: FWIW, I'd rather just inline it and avoid maintaining APIs that are just for tests in our production code.

**Review by vvcephei — APPROVED** [review]
Thanks, @abbccdda !

One small thought, which you can take or leave as you see fit.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:739]: I see your point and thought about it, but I feel a little bit reluctant to remove it for the reason that it is indeed a handy call to make for testing purpose :)

**Comment by mjsax** [comment]
Retest this please.

**Comment by ableegoldman** [comment]
What about all the other times we commit tasks in TaskManager? Aren't we still vulnerable to committing a CREATED task during `handleAssignment`, for example? Imho it makes sense to push this logic into the StreamTask/StandbyTask rather than chase down all code paths that may commit and make sure we are passing in only "valid" tasks

**Comment by mjsax** [comment]
Java 8 passed.
Java 11 and 14 failed. (Test results are not longer available.)

@ableegoldman You raise a good point. Can we revisit this when doing the kip-447 follow up refactoring? I would like to merge this PR to we can re-deploy soak with the fix?

**Comment by mjsax** [comment]
Retest this please

**Comment by ableegoldman** [comment]
Sure thing, that doesn't need to block this PR. As long as it doesn't slip off the radar completely I'm 👍

**Comment by abbccdda** [comment]
Got 2/3 green build with only one test failure:
```
org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[false]

java.lang.AssertionError: Condition not met within timeout 60000. Clients did not startup and stabilize on time. Observed transitions: 
  client-1 transitions: [KeyValue(RUNNING, REBALANCING), KeyValue(REBALANCING, RUNNING)]
  client-2 transitions: [KeyValue(CREATED, REBALANCING), KeyValue(REBALANCING, RUNNING), KeyValue(RUNNING, REBALANCING), KeyValue(REBALANCING, RUNNING)]
```

**Review by guozhangwang — APPROVED** [review]
LGTM! I left a comment on JIRA ticket as well regarding in the future, if we can make the state of the individual task to NOT be exposed into the task-manager, but for now I think having it inside task-manager is okay.

**Comment by Aatirhassanpir** [comment]
Title
KAFKA-9972: Only commit tasks with valid states

Repository Version
Apache Kafka Streams (version unknown, as PR details don’t specify)

Key Words / Search Terms
Task commit, invalid state, CREATED state, IllegalStateException, task validation, TaskManager, Kafka Streams

Description
During a soak test, it was observed that a standby task could enter the CREATED state during a commit operation. This resulted in an IllegalStateException, disrupting normal processing and causing runtime instability. The issue arises because the TaskManager does not currently validate the state of tasks before committing them.
To resolve this issue, a state validation check must be implemented in the TaskManager. This check ensures that only tasks in valid states are committed, thereby preventing runtime errors caused by committing tasks in improper states like CREATED.

Expected Behavior
The TaskManager should commit only tasks that are in valid states.
Invalid states such as CREATED should be skipped, preventing the IllegalStateException.

Current Behavior
The TaskManager attempts to commit tasks without validating their state.
Tasks in invalid states, such as CREATED, trigger an IllegalStateException, causing processing disruptions.

Steps to reproduce (for bug reports)
Simulate a scenario where a standby task enters the CREATED state during a commit operation.
Attempt to commit tasks using the current implementation of TaskManager.
Observe the IllegalStateException triggered by the invalid state.
Use cases / motivating examples (for feature requests)
This fix ensures stable and reliable operation of Kafka Streams by:

Preventing runtime errors caused by committing tasks in invalid states.
Enabling smoother task handling during high-throughput operations.

**Comment by mjsax** [comment]
@Aatirhassanpir Not sure what you comment is about. But if you want to report a bug, please file a Jira ticket. Adding a comment to a PR does not allow to track us any bug report.

**Pull request merged**
